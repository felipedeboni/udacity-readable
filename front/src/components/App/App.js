import React from 'react';
import { Switch, Route } from 'react-router-dom';
import 'components/App/App.css';
import {
  CommentFormContainer,
  PostContainer,
  PostsListContainer,
  PostFormContainer
} from 'containers';

import { Header, Post, PageNotFound } from 'components';

const App = ({ categories }) => {
  const renderRouteOrNotFound = (props, component) => {
    const { match } = props;
    const categoriesArr = categories.map(_ => _.path);
    const matcher = new RegExp(`^(${categoriesArr.join('|')})`);
    if (matcher.test(match.params.category)) {
      return React.createElement(component, {
        ...props
      });
    } else {
      return <PageNotFound />;
    }
  };

  return categories ? (
    <div className="app container-fluid">
      <Header categories={categories} />
      <Switch>
        <Route
          exact
          path="/"
          children={props => <PostsListContainer category="" {...props} />}
        />
        <Route
          exact
          path="/posts/new"
          children={props => (
            <PostFormContainer categories={categories} {...props} />
          )}
        />
        <Route
          exact
          path="/comments/:comment/edit"
          render={_props => (
            <CommentFormContainer
              id={_props.match.params.comment}
              {..._props}
            />
          )}
        />
        <Route
          path="/:category"
          exact
          strict
          render={_props => {
            const props = {
              ..._props,
              category: _props.match.params.category
            };
            return renderRouteOrNotFound(props, PostsListContainer);
          }}
        />
        <Route
          exact
          path="/:category/:post"
          render={_props => {
            const props = {
              ..._props,
              postId: _props.match.params.post,
              component: Post
            };
            return renderRouteOrNotFound(props, PostContainer);
          }}
        />
        <Route
          exact
          path="/:category/:post/edit"
          render={_props => {
            const props = {
              ..._props,
              categories,
              id: _props.match.params.post
            };
            return renderRouteOrNotFound(props, PostFormContainer);
          }}
        />
        <Route path="*" component={PageNotFound} />
      </Switch>
    </div>
  ) : (
    ''
  );
}

export default App;
