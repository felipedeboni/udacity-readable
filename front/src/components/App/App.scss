.app {
  max-width: 990px;
}
