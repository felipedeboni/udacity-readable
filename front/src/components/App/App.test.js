import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

const categoriesMock = [
  {
    name: 'react',
    path: 'react'
  },
  {
    name: 'redux',
    path: 'redux'
  },
  {
    name: 'udacity',
    path: 'udacity'
  }
];

describe('App', () => {
  it('renders without crashing', () => {
    const test = <App />;

    expect(shallow(test)).toMatchSnapshot();
  });

  it('renders with categories', () => {
    const test = <App categories={categoriesMock} />;

    expect(shallow(test)).toMatchSnapshot();
  });
});
