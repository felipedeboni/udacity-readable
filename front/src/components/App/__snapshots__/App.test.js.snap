// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders with categories 1`] = `
<div
  className="app container-fluid"
>
  <Header
    categories={
      Array [
        Object {
          "name": "react",
          "path": "react",
        },
        Object {
          "name": "redux",
          "path": "redux",
        },
        Object {
          "name": "udacity",
          "path": "udacity",
        },
      ]
    }
  />
  <Switch>
    <Route
      exact={true}
      path="/"
    />
    <Route
      exact={true}
      path="/posts/new"
    />
    <Route
      exact={true}
      path="/comments/:comment/edit"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/:category"
      render={[Function]}
      strict={true}
    />
    <Route
      exact={true}
      path="/:category/:post"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/:category/:post/edit"
      render={[Function]}
    />
    <Route
      component={[Function]}
      path="*"
    />
  </Switch>
</div>
`;

exports[`App renders without crashing 1`] = `""`;
